vochain/processarchive/p2p.go

```diff
@@ -0,0 +1,87 @@
+package processarchive
+
+import (
+	"context"
+	"encoding/base64"
+	"time"
+
+	ipfscrypto "github.com/libp2p/go-libp2p-core/crypto"
+	"go.vocdoni.io/dvote/data"
+	"go.vocdoni.io/dvote/log"
+)
+
+const (
+	publishInterval = 10 * time.Minute
+	ipnsKeyAlias    = "processarchivekey"
+)
+
+// AddKey adds a base64 encoded ECDSA 256bit private key or generates
+// a new one.
+func (p *ProcessArchive) AddKey(b64key string) error {
+	var ipnsPk []byte
+	var err error
+	if b64key == "" {
+		// if key already exist, just return
+		if _, err := p.GetKey(); err == nil {
+			return nil
+		}
+		// else generate a new key
+		ipnsPk = data.NewIPFSkey()
+	} else {
+		ipnsPk, err = base64.StdEncoding.DecodeString(b64key)
+		if err != nil {
+			return err
+		}
+	}
+	return p.ipfs.AddKeyToKeystore(ipnsKeyAlias, ipnsPk)
+}
+
+// GetKey fetch the base64 encoded IPFS private key used to
+// publish the IPNS record.
+func (p *ProcessArchive) GetKey() (string, error) {
+	pk, err := p.ipfs.Node.Repo.Keystore().Get(ipnsKeyAlias)
+	if err != nil {
+		return "", err
+	}
+	pkb, err := ipfscrypto.MarshalPrivateKey(pk)
+	if err != nil {
+		return "", err
+	}
+	return base64.StdEncoding.EncodeToString(pkb), nil
+}
+
+// Publish is a blocking routine that reads ProcessArchive.publish channel
+// in order to trigger a new IPNS record announcement publishing the
+// process archive directory.
+func (p *ProcessArchive) Publish() {
+	log.Infof("starting process archive IPNS publish daemon with interval %s",
+		publishInterval)
+	// Wait for publishInterval and ensure a first execution
+	p.lastUpdate = time.Now().Add(-publishInterval)
+	p.publish <- true
+	select {
+	case <-p.publish:
+		if time.Since(p.lastUpdate) < publishInterval {
+			break
+		}
+		p.lastUpdate = time.Now()
+		// make it async so the channel does not get full and blocking
+		go func() {
+			p.publishLock.Lock()
+			defer p.publishLock.Unlock()
```

Comment:
I don't get the point of this lock. If two goroutines call Publish concurrently, this func is racy because of how we use lastUpdate. And if we don't support calling Publish concurrently, then this lock is unnecessary.

Reply:
Its for safety. There might be some corner case where a new iteration happens while the current one has not yet finished.
I'm trying to ensure this does not happend by using the context.WithTimeout but some time adjustments on the host computer or something else might broke this logic. So that lock just tries to handle kind of safe this very unlikely situation.

Reply:
I still don't get it though. This doesn't run in a loop nor is it called many times. And, as said above, if concurrent calls to this func are happening, then the lock isn't saving you - the func is still racy.